boutdata/restart.py

```diff
@@ -956,3 +956,213 @@ def addvar(var, value, path="."):
 
             # Set the variable in the NetCDF file
             df.write(var, data)
+
+
+def change_grid(
+    from_grid_file,
+    to_grid_file,
+    path="data",
+    output=".",
+    interpolator="nearest",
+    show=False,
+):
+    """
+    Convert a set of restart files from one grid to another
+
+    Notes:
+    - Only working for 2D (axisymmetric) simulations with nz = 1
+
+    from_grid_file : str
+         File containing the input grid
+    to_grid_file : str
+         File containing the output grid
+    path : str, optional
+         Directory containing input restart files
+    output : str, optional
+         Directory where output restart files will be written
+    interpolator : str, optional
+         Interpolation method to use. Options are 'nearest', 'CloughTocher', 'RBF'
+    show : bool, optional
+         Display the interpolated fields using Matplotlib
+
+    """
+
+    # Read in grid files
+    with DataFile(from_grid_file) as g:
+        from_Rxy = g["Rxy"]
+        from_Zxy = g["Zxy"]
+
+    with DataFile(to_grid_file) as g:
+        to_Rxy = g["Rxy"]
+        to_Zxy = g["Zxy"]
+
+    file_list = glob.glob(os.path.join(path, "BOUT.restart.*.nc"))
+    if len(file_list) == 0:
+        raise ValueError("ERROR: No restart files found")
+
+    copy_vars = [
+        "NXPE",
+        "NYPE",
+        "hist_hi",
+        "tt",
+        "MXG",
+        "MYG",
+        "MZG",
+        "nz",
+        "run_id",
+        "run_restart_from",
+    ]
+    copy_data = {}
+    interp_vars = []
+
+    # Read information from a restart file
+    with DataFile(file_list[0]) as f:
+        for var in copy_vars:
+            copy_data[var] = f[var]
+
+        # Get a list of variables
+        varnames = f.list()
+
+        for var in varnames:
+            dimensions = f.dimensions(var)
+            if dimensions == ("x", "y", "z"):
+                # Could be an evolving variable [x,y,z]x
+                interp_vars.append(var)
+
+    # Only tested for nz = 1
+    assert copy_data["nz"] == 1
+
+    def extrapolate_yguards(data2d, myg):
+        nx, ny = data2d.shape
+        result = np.zeros((nx, ny + 2 * myg))
+        print(result.shape)
+        result[:, myg:-myg] = data2d
+
+        dy = result[:, myg + 1] - result[:, myg]
+        for i in range(1, myg + 1):
+            result[:, myg - i] = result[:, myg] - i * dy
+        dy = result[:, -myg - 1] - result[:, -myg - 2]
+        for i in range(1, myg + 1):
+            result[:, -myg - 1 + i] = result[:, -myg - 1] + i * dy
+        return result
+
+    from_Rxy = extrapolate_yguards(from_Rxy, copy_data["MYG"])
+    from_Zxy = extrapolate_yguards(from_Zxy, copy_data["MYG"])
+    to_Rxy = extrapolate_yguards(to_Rxy, copy_data["MYG"])
+    to_Zxy = extrapolate_yguards(to_Zxy, copy_data["MYG"])
```

Comment:
What if the grid file contains y-boundary cells? It gets a bit complicated to deal with, especially allowing for an upper divertor, so suggest just adding some check like
```
if g.get("y_boundary_guards", 0) != 0:
    raise ValueError("Support for grid files with y-boundary cells not implemented yet")
```
when opening the grid files.

Reply:
Thanks @johnomotani for the suggestions, and happy New Year!